4.preliminary_results/scripts/0.UMAP_embeddings.py

```diff
@@ -0,0 +1,210 @@
+        cell_line_data = pd.read_parquet(
+            file_path,
+            filters=[("Metadata_cell_line", "==", cell_line)]
+        )
+        
+        # perform stratified sampling within the cell line
+        stratified_samples = cell_line_data.groupby("Metadata_seeding_density", group_keys=False).apply(
+            lambda group: group.sample(
+                n=min(samples_per_cell_line // len(cell_line_data["Metadata_seeding_density"].unique()), len(group)),
+                random_state=random_seed
+            ),
+            include_groups=False  # Explicitly exclude grouping columns in the output
+        ).reset_index(drop=True)
+        
+        # add the stratified sampled data to list
+        all_samples.append(stratified_samples)
+    
+    # combine all samples for the current plate
+    combined_samples = pd.concat(all_samples, ignore_index=True)
+    
+    # adjust total number of samples
+    sampled_df = combined_samples.sample(
+        n=min(total_samples, len(combined_samples)),
+        random_state=random_seed,
+        replace=False
+    )
+    
+    # store results
+    sampled_data_dict[plate_name] = sampled_df
+    print(f"Processed plate: {plate_name} | Sampled data shape: {sampled_df.shape}")
+
+
+# ## Generate UMAP embeddings per plate
+
+# In[4]:
+
+
+# UMAP configuration
+umap_components = 2
+random_seed = 0
+
+# dictionary to store UMAP results
+umap_results_dict = {}
+
+# process sampled data from the sampled_data dictionary
+for plate_name, sampled_df in sampled_data_dict.items():
+    # separate metadata and feature columns
+    metadata_columns = [col for col in sampled_df.columns if col.startswith("Metadata_")]
+    feature_columns = [col for col in sampled_df.columns if not col.startswith("Metadata_")]
+    
+    # drop rows with NaN values in feature columns
+    cleaned_df = sampled_df.dropna(subset=feature_columns)
+    
+    # perform UMAP embedding on the cleaned feature data
+    umap_model = umap.UMAP(n_components=umap_components, random_state=random_seed, n_jobs=1)
+    umap_embeddings = umap_model.fit_transform(cleaned_df[feature_columns])
+    
+    # create a DataFrame for embeddings
+    umap_df = pd.DataFrame(
+        umap_embeddings,
+        columns=[f"UMAP{i}" for i in range(umap_components)],
+        index=cleaned_df.index
+    )
+    
+    # combine UMAP embeddings with metadata
+    final_df = pd.concat([cleaned_df[metadata_columns], umap_df], axis=1)
+    
+    # store the result in the dictionary to use for outputting results
+    umap_results_dict[plate_name] = final_df
+    
+    print(f"UMAP embeddings generated for plate: {plate_name}")
+    print(f"Cleaned samples shape: {cleaned_df.shape}")
+    print(f"Final shape with embeddings: {final_df.shape}")
+
+
+# ## Create UMAP plots per plate labelling by the cell line
+
+# In[5]:
+
+
+# Define consistent color for U2-OS
+u2os_color = "#9b0068"  # Darker magenta color
+custom_palette = {}
+
```

Comment:
Interesting to see UMAPs with this!

Reply:
Thanks! I hope that it will foster some discussion, I think I want to try it with the bulk data next!